src/include/server/mir/scene/text_input_hub.h

```diff
@@ -97,6 +98,48 @@ struct TextInputState
     std::optional<TextInputContentPurpose> content_purpose;
 };
 
+class CopyableWlArray
+{
+public:
+    explicit CopyableWlArray(wl_array* in_data)
+        : data_{copy(in_data)}
+    {
+    }
+
+    CopyableWlArray(CopyableWlArray const& other)
+        : data_{copy(other.data_)}
+    {
+    }
+
+    CopyableWlArray& operator=(CopyableWlArray const& other)
+    {
+        data_ = copy(other.data_);
+        return *this;
+    }
```

Comment:
This leaks `data_`. The canonical form of assignment is:
```suggestion
    CopyableWlArray& operator=(CopyableWlArray other)
    {
        std::swap(data, other.data);
        return *this;
    }
```